File: Sources/PDFjet/Annotation.swift
/**
 * Annotation.swift
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */

///
/// Used to create PDF annotation objects.
///
// Annotation.swift
// Used to create PDF annotation objects.
class Annotation {
    static let Link = "Link"
    static let FileAttachment = "FileAttachment"
    static let Polygon = "Polygon"
    static let Circle = "Circle"
    static let Square = "Square"
    static let Text = "Text"

    var objNumber: Int = 0
    var annotationType: String?
    var x1: Float = 0.0
    var y1: Float = 0.0
    var x2: Float = 0.0
    var y2: Float = 0.0
    var vertices: [Float]?
    var fillColor: [Float]?
    var transparency: Float = 0.0
    var title: String?
    var contents: String?
    var uri: String?
    var key: String?
    var language: String?
    var actualText: String?
    var altDescription: String?
    var fileAttachment: FileAttachment?
    // Set once the annotation has been written with a /StructParent key.
    var structParentWritten = false

    /// Creates an annotation object.
    /// - Parameters:
    ///   - annotationType: The annotation type.
    ///   - x1: The x coordinate of the top left corner.
    ///   - y1: The y coordinate of the top left corner.
    ///   - x2: The x coordinate of the bottom right corner.
    ///   - y2: The y coordinate of the bottom right corner.
    ///   - vertices: The polygon annotation vertices.
    ///   - fillColor: The fill color as RGB floats.
    ///   - transparency: The transparency value (0.0 to 1.0).
    ///   - title: The annotation title.
    ///   - contents: The annotation content/description.
    ///   - uri: The URI string.
    ///   - key: The destination name.
    ///   - language: The language code.
    ///   - actualText: The actual text content. Defaults to uri if nil.
    ///   - altDescription: Alternative description. Defaults to uri if nil.
    init(
        _ annotationType: String?,
        _ x1: Float,
        _ y1: Float,
        _ x2: Float,
        _ y2: Float,
        _ vertices: [Float]?,
        _ fillColor: [Float]?,
        _ transparency: Float,
        _ title: String?,
        _ contents: String?,
        _ uri: String?,
        _ key: String?,
        _ language: String?,
        _ actualText: String?,
        _ altDescription: String?
    ) {
        self.annotationType = annotationType
        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2
        self.vertices = vertices
        self.fillColor = fillColor
        self.transparency = transparency
        self.title = title
        self.contents = contents
        self.uri = uri
        self.key = key
        self.language = language

        // Match Java ternary logic: (actualText == null) ? uri : actualText
        // A link created from a destination name has no uri to fall back on,
        // so use the name itself rather than leaving the link undescribed.
        let fallback = uri ?? key
        self.actualText = actualText ?? fallback
        self.altDescription = altDescription ?? fallback

        self.fileAttachment = nil // Will be set externally if needed
    }
}

File: com/pdfjet/Annotation.java
/*
 * Annotation.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet;

/**
 * Used to create PDF annotation objects.
 */
class Annotation {
    public static final String Link = "Link";
    public static final String FileAttachment = "FileAttachment";
    public static final String Polygon = "Polygon";
    public static final String Circle = "Circle";
    public static final String Square = "Square";
    public static final String Text = "Text";

    int objNumber;
    String annotationType = null;
    float x1 = 0f;
    float y1 = 0f;
    float x2 = 0f;
    float y2 = 0f;
    float[] vertices = null;
    float[] fillColor = null;
    float transparency = 0f;
    String title = null;
    String contents = null;
    String uri = null;
    String key = null;
    String language = null;
    String actualText = null;
    String altDescription = null;
    FileAttachment fileAttachment = null;
    // Set once the annotation has been written with a /StructParent key.
    boolean structParentWritten = false;

    /**
     *  This class is used to create annotation objects.
     *
     *  @param annotationType the annotation type.
     *  @param x1 the x coordinate of the top left corner.
     *  @param y1 the y coordinate of the top left corner.
     *  @param x2 the x coordinate of the bottom right corner.
     *  @param y2 the y coordinate of the bottom right corner.
     *  @param vertices the polygon annotation vertices.
     *  @param uri the URI string.
     *  @param key the destination name.
     */
    Annotation(
            String annotationType,
            float x1,
            float y1,
            float x2,
            float y2,
            float[] vertices,
            float[] fillColor,
            float transparency,
            String title,
            String contents,
            String uri,
            String key,
            String language,
            String actualText,
            String altDescription) {
        this.annotationType = annotationType;
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
        this.vertices = vertices;
        this.fillColor = fillColor;
        this.transparency = transparency;
        this.title = title;
        this.contents = contents;
        this.uri = uri;
        this.key = key;
        this.language = language;
        // A link created from a destination name has no uri to fall back on,
        // so use the name itself rather than leaving the link undescribed.
        String fallback = (uri != null) ? uri : key;
        this.actualText = (actualText == null) ? fallback : actualText;
        this.altDescription = (altDescription == null) ? fallback : altDescription;
    }
}
